### PR TITLE
[BREAKING] feat: add pallet-proxy to all runtimes

### PR DESCRIPTION
**Pull request by gregdhill**
Signed-off-by: Gregory Hill <gregorydhill@outlook.com>

This has been requested by several users. The [proxy pallet](https://github.com/paritytech/substrate/tree/master/frame/proxy) allows accounts to dispatch calls on behalf of another account (with their permission).

### CODE REVIEW

**Comment by gregdhill** [comment]
Before we merge this, we should consider the following proxy types (requested):
- claiming rewards
- transfers
- issues / redeems

**Comment by sander2** [comment]
> Before we merge this, we should consider the following proxy types (requested):
> 
>     * claiming rewards
> 
>     * transfers
> 
>     * issues / redeems

This seems like something that you should discuss with @alexbleau or @nud3l

**Comment by nud3l** [comment]
Do we have to decide in advance which pallets we support for proxy?

**Comment by sander2** [comment]
> Do we have to decide in advance which pallets we support for proxy?

If we can decide now, that'd be ideal. Of course, we can always do a runtime upgrade to make changes. It would be a breaking change because of changing metadata, but it's not the end of the world. In the current version, you can allow other accounts to create _any_ call on behalf of you, which might find some usage already

**Comment by nud3l** [comment]
Happy to have this allow any proxies for the first version.

File: parachain/runtime/interlay/src/lib.rs [hunk 1]
```diff
@@ -992,6 +992,73 @@ impl pallet_identity::Config for Runtime {
...
+impl Default for ProxyType {
+    fn default() -> Self {
+        Self::Any
+    }
+}
+
+impl InstanceFilter<RuntimeCall> for ProxyType {
+    fn filter(&self, c: &RuntimeCall) -> bool {
+        match self {
+            _ if matches!(c, RuntimeCall::Utility(..)) => true,
```
Comment (sander2) [inline parachain/runtime/interlay/src/lib.rs:1033]: The purpose of this isn't clear to me. It seems to have no use, can we remove?

File: parachain/runtime/interlay/src/lib.rs [hunk 2]
```diff
@@ -992,6 +992,73 @@ impl pallet_identity::Config for Runtime {
...
+            ProxyType::Any => true,
+        }
+    }
+    fn is_superset(&self, o: &Self) -> bool {
+        match (self, o) {
+            (x, y) if x == y => true,
+            (ProxyType::Any, _) => true,
+            (_, ProxyType::Any) => false,
+            _ => false,
```
Comment (sander2) [inline parachain/runtime/interlay/src/lib.rs:1042]: please fix warnings

File: parachain/runtime/testnet-kintsugi/src/lib.rs [hunk 3]
```diff
@@ -1051,13 +1118,14 @@ construct_runtime! {
...
         System: frame_system::{Pallet, Call, Storage, Config, Event<T>} = 0,
         Timestamp: pallet_timestamp::{Pallet, Call, Storage, Inherent} = 1,
-        Sudo: pallet_sudo::{Pallet, Call, Storage, Config<T>, Event<T>} = 2,
-        Utility: pallet_utility::{Pallet, Call, Event} = 3,
-        TransactionPayment: pallet_transaction_payment::{Pallet, Storage, Event<T>} = 4,
-        Scheduler: pallet_scheduler::{Pallet, Call, Storage, Event<T>} = 5,
-        Preimage: pallet_preimage::{Pallet, Call, Storage, Event<T>} = 6,
-        Multisig: pallet_multisig::{Pallet, Call, Storage, Event<T>} = 7,
-        Identity: pallet_identity::{Pallet, Call, Storage, Event<T>} = 8,
+        Utility: pallet_utility::{Pallet, Call, Event} = 2,
```
Comment (sander2) [inline parachain/runtime/testnet-kintsugi/src/lib.rs:1121]: Why this change? If you really want this, it'll be a breaking change. Also, please mark the pr as breaking

File: parachain/runtime/interlay/src/lib.rs (hunk 1 above)
Comment (gregdhill) [reply parachain/runtime/interlay/src/lib.rs:1033]: See this comment: https://github.com/AcalaNetwork/Acala/blob/a0251084b49b40d69262c84d13f4f9ed86729f1d/runtime/karura/src/lib.rs#L1291-L1293

File: parachain/runtime/testnet-kintsugi/src/lib.rs (hunk 3 above)
Comment (gregdhill) [reply parachain/runtime/testnet-kintsugi/src/lib.rs:1121]: To match the Kintsugi runtime, otherwise the testnet will have different indices

Comment (gregdhill) [reply parachain/runtime/testnet-kintsugi/src/lib.rs:1120]: Do we want to consider these breaking changes also to the testnet runtimes?

File: parachain/runtime/interlay/src/lib.rs (hunk 1 above)
Comment (gregdhill) [reply parachain/runtime/interlay/src/lib.rs:1034]: It is possible that this doesn't apply to `ProxyType::Any` but since in that case this is anyway a no-op, perhaps we should leave this so it is not forgotten in future updates?

File: parachain/runtime/testnet-kintsugi/src/lib.rs (hunk 3 above)
Comment (sander2) [reply parachain/runtime/testnet-kintsugi/src/lib.rs:1120]: I'm not sure, but this pr is breaking regardless (assuming the `proxy` pallet has events)

Comment (gregdhill) [reply parachain/runtime/testnet-kintsugi/src/lib.rs:1120]: Updated the title

File: parachain/runtime/interlay/src/lib.rs (hunk 1 above)
Comment (gregdhill) [reply parachain/runtime/interlay/src/lib.rs:1034]: Okay I did some investigating, the way it works is if a `ProxyDefinition` exists it adds filters to the origin of the call: https://github.com/paritytech/substrate/blob/946507ba9ef13e263534176b7b74e26fc56efbd4/frame/proxy/src/lib.rs#L784

Say we dispatch `Utility::batch(Tokens::transfer(..))` (using `ProxyType::Any`) via the proxy pallet this will re-dispatch [here](https://github.com/paritytech/substrate/blob/946507ba9ef13e263534176b7b74e26fc56efbd4/frame/utility/src/lib.rs#L224) due to the above rule but then the inner call will be reevaluated using the other filter rules. So if we only support the `Any` permission I think we can drop the `Utility` filter.

Comment (sander2) [reply parachain/runtime/interlay/src/lib.rs]: Ah ok, and then and calls inside a batch get dispatched with that additional filter. Ok that makes sense, thanks for looking into it

**Review by sander2 — APPROVED** [review]
